### ISSUE TITLE
Unable to change font to "Segoe MDL2 Assets"

### ISSUE THREAD

**Issue by runiter** [JFXtras/jfxtras-styles#138]
I recently upgraded my jmetro jar file from v5.4 to v8.6.5
But the font seems to be broken now.

I tried both css:

```
.toolbarButton{
    -fx-font-family : "Segoe MDL2 Assets";
}
```

And java code:

```
button.setFont(Font.font("Segoe MDL2 Assets"));
button.setText("\uE74E");
```

Neither works anymore.

**Comment by dukke** [comment]
Hi Saeid!

This is probably not an issue in JMetro, more how JavaFX CSS works (which actually works a lot like web CSS). JMetro defines rules in css but doesn't override how the JavaFX CSS mechanisms work (which rules override which, which stylesheets have higher priority, etc).

A few things you need to understand about JavaFX CSS:

- First JMetro is different from Modena, JMetro is either a Scene or Parent stylesheet whereas Modena is a user agent stylesheet. This means that JMetro has the same priority as the stylesheets you provide to it. (BTW: All JavaFX themes I know work this way, however I'll try to have JMetro be a user agent stylesheet. Which is probably better. There's already an issue for that).
To override the rules in JMetro stylesheets you need to define rules with the same or with a higher level of specificity. I say the same level of specificity will also work, because if 2 rules have the same level, the last one will win.
Google CSS specificity (web css is similar to JavaFX CSS so if you see articles on web CSS you can also learn from it with the necessary adaptations)
- JMetro CSS and your scene/parent CSS will always override the same properties in code (and also FXML which ends up being code). Only inline styles have higher priority than scene/parent stylesheets.
- You need to import the new font in your CSS stylesheet (you might be doing this already, but just pointing this out to be sure)

A reference document you should also check out (if you haven't already): https://docs.oracle.com/javase/8/javafx/api/javafx/scene/doc-files/cssref.html#typefont

Thanks

**Comment by runiter** [comment]
Thanks for your reply.
Two observation:

1. You said this is probably not an issue with JMetro. But this used to work before I updated JMetro. Even right now when I downgrade JMetro to v5.4 it works again.

2. If I explicitly call `button.setFont` method 10 seconds after the scene was created and shown with JMetro skin, shouldn't it override whatever font was set by JMetro?

**Comment by dukke** [comment]
1 - Rules might have changed or might have been added. There might be a conflict between your rule and JMetro and given JMetro is more specific it wins. Just try to make your rule as specific as possible and see if it works. More specific than JMetro.

2 - No. Scene or Parent stylesheets always override properties set in code. Only inline styles override scene/parent stylesheets.. On that website url I posted, there is a section that talks about the priority of the various types of rules (user agent stylesheet, stylesheets, code, etc).

Thanks!

**Comment by runiter** [comment]
Thank you. Makes sense.

Can you tell me how to define rules with the same or with a higher level of specificity?

This is what I have currently:

gui.css:
```
.symbolButton{
    -fx-font-family : "Segoe MDL2 Assets";
    -fx-font-size: 15;
    -fx-padding: 3 2 2 2;
}
```
MainPanel.fxml:
```
<VBox stylesheets="@gui.css" xmlns="http://javafx.com/javafx/8.0.171" xmlns:fx="http://javafx.com/fxml/1">
   ...
   <Button styleClass="toolbarButton" text="Random Text">
   </Button>
```
Java:
```
FXMLLoader loader = new FXMLLoader(MainPanel.class.getResource("MainPanel.fxml"));
Parent root = loader.load();
Scene scene = new Scene(root, 1600, 1000);
scene.getStylesheets().add("runiter/calculator/gui/gui.css");
JMetro jMetro = new JMetro(scene, Style.DARK);
```

**Comment by dukke** [comment]
Hi again Saeid,

Yeah that has a very low specificity, i.e. there is only one styleclass in that rule.

Google CSS specificity. Not sure there are many articles for JavaFX CSS but it works very similar to Web CSS.

**Comment by runiter** [comment]
Thanks I now understand CSS specificity.
But that's not helping me figure out how to change the font of a Javafx Button!

Could you please tell me how to change my css code to make it more specific?

**Comment by dukke** [comment]
Hi Saeid,

I hope you understand, but given how much limited time I have and being this an hobby project I don't have time to end up answering JavaFX questions. Please understand. All this time I spend developing JMetro is never payed and very little recognized, and sometimes I question if it's worth it. It would be too much for me if I end up answering general JavaFX questions as well.

There are other forums and places that are better for this types of questions, for example StackOverflow.

I'll just give you this one advice: add more styleclasses, IDs, etc to your rule (IDs have a greater value for specificity than styleclasses). Like this:
```
.button.symbolButton {
  -fx-font-family : "Segoe MDL2 Assets";
}
```
This has 2 styleclasses so it's more valuable than the definition in JMetro.

Also, like I said, and this is probably the issue, make sure you are importing that font in your CSS file (consult the link I posted). I don't see any of that in the code you posted. 

Thanks.

**Comment by runiter** [comment]
I understand. Normally I would post this kind of question in StackOverflow but since this issue is not present in Javafx Modena I doubt I could get a useful response.

I hope you won't give up on this project. It's a fantastic project that I think is badly needed. In fact I think this should be part of Javafx standard and Windows 10 app should default to this font.

I don't think I need to import this font because this font comes standard with Windows 10 and also the follow inline actually works fine:

`<Button style="-fx-font-family: &quot;Segoe MDL2 Assets&quot;;" text="My Symbol">`

I tried other fonts like "Times New Roman" and same problem.

I guess I will just use the inline styling for now.
But I think other users of your library will struggle with this issue too.

**Comment by dukke** [comment]
Hi again Saeid,

Thanks for the kind words. Appreciate it. :)
Also thanks for registering issues, some of them have contributed highly to improving JMetro.

Seems like it's pretty probable that it is a specificity issue (i.e. your rule not beating the specificity of another rule already defined, probably in JMetro). That's why inline styles work, since they have higher priority than stylesheet styles.

Have you tried my suggestion? Also one other thing you could do, is use IDs since IDs contribute more points towards specificity than styleclasses. Use has many styleclasses, IDs, etc as possible in that rule, by specifying parent, grand-parent, etc of that button. Also direct child selectors contributes more points than the "any child" (space) selector. Also check JMetro base.css and make sure you are overriding any of the possible rules being applied from that stylesheet.

I've seen people struggle with this before as well. To my understanding the problem is this:
1 - Please don't that this comment personally, I've also seen other people make this same mistake. I think people don't understand CSS enough and all it's particularities. Usually they assume they do and that they don't need to study it, and end up not understanding why things don't end up working and assume it's some bug or problem with JavaFX or the third party library. I've also seen this kind of mistake happening with developers working with web CSS. Study specificity thoroughly.
2 - There's no adequate JavaFX tools to debug this. In the web there is firebug, web inspector, etc. They'll show you which rules are being overriden, which ones are being applied and where in the files they are specified.
3 - JMetro isn't a user agent stylesheet like Modena. People assume it is but it isn't. And this has big implications on how you need to specify your own stylesheets.

My objective in the near future, will be tackling (3). There's already an issue registered for that. I'll have to convert JMetro to being an user agent stylesheet and so it'll work like Modena. And for (1) perhaps create a blog post.
(2) would require a better specialized JavaFX tool which would need more time to develop which I don't have time to at the moment. Would be a good thing for Oracle, Gluon to do IMO.

Thanks, hope this helps.

**Comment by runiter** [comment]
Hi Pedro,
I tried your suggestion and it didn't work.
But then I found this useful tool for debugging Javafx scenes:

https://github.com/JonathanGiles/scenic-view

It helped me figure out that there is an extra "text" component inside the Button and to be more specific I need to target that component directly instead. So the following finally worked:

```
.symbolButton .text{
    -fx-font-family : "Segoe MDL2 Assets";
}
```

Yay!

I total agree with your (3) proposal. I think it would help the popularity of JMetro if users could switch from Modena to JMetro without the need of editing their css files. This could be a huge barrier to mass adaption of your otherwise superior library.
